### PR TITLE
Use WaitingTaskHolder for function arguments

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Switched from WaitingTask* to WaitingTaskHolder in framework APIs. 
- This increases thread safety (by avoiding accidental starvation if a task is not run) and will make it easier to transition away from the deprecated TBB APIs.

#### PR validation:

The code compiles and all framework related unit tests pass.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Framework/src/DataProxy.cc [hunk 1]
```diff
@@ -102,7 +103,7 @@ namespace edm {
         auto waitTaskPtr = waitTask.get();
         auto token = ServiceRegistry::instance().presentToken();
         edm::esTaskArena().execute([this, waitTaskPtr, &iRecord, &iKey, iEventSetupImpl, token]() {
-          prefetchAsync(waitTaskPtr, iRecord, iKey, iEventSetupImpl, token);
+          { prefetchAsync(WaitingTaskHolder(waitTaskPtr), iRecord, iKey, iEventSetupImpl, token); }
```
Comment (makortel) [inline FWCore/Framework/src/DataProxy.cc:106]: Why the extra braces?

File: FWCore/Framework/src/Worker.h [hunk 2]
```diff
@@ -984,9 +980,12 @@ namespace edm {
         auto ownRunTask = std::make_shared<DestroyTask>(runTask);
         auto selectionTask = make_waiting_task(
             tbb::task::allocate_root(),
-            [ownRunTask, parentContext, info = transitionInfo, token, this](std::exception_ptr const*) mutable {
-              ServiceRegistry::Operate guard(token);
-              prefetchAsync<T>(ownRunTask->release(), token, parentContext, info, T::transition_);
+            [ownRunTask, parentContext, info = transitionInfo, token, this](std::exception_ptr const* iExcept) mutable {
+              //NOTE: a non null value of iExcept means the selection succeeded
+              if (!iExcept) {
```
Comment (makortel) [inline FWCore/Framework/src/Worker.h:985]: Could you elaborate? Was there a bug before?

File: FWCore/Framework/src/DataProxy.cc (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Framework/src/DataProxy.cc:106]: I think in an earlier version of my edit I had a WaitingTaskHolder variable so the scope was needed. Now they are just a left over.

File: FWCore/Framework/src/Worker.h (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Framework/src/Worker.h:985]: That is part of a change I tried to back out. Evidently I missed a piece.

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Framework/src/Worker.cc [hunk 3]
```diff
@@ -267,63 +268,65 @@ namespace edm {
...
+                for (size_t i = 0; i != items.size(); ++i) {
+                  if (recs[i] != ESRecordIndex{}) {
+                    auto rec = iImpl.findImpl(recs[i]);
+                    if (rec) {
+                      rec->prefetchAsync(hWaitTask, items[i], &iImpl, iToken);
+                    }
+                  }
+                }
+              }
+              waitTask->decrement_ref_count();
```
Comment (makortel) [inline FWCore/Framework/src/Worker.cc:290]: How useful is the
```cpp
waitTask->set_ref_count(2);
{ ... }
waitTask->decrement_ref_count();
waitTask->wait_for_all();
```
compared to
```cpp
waitTask->set_ref_count(1);
{ ... }
waitTask->wait_for_all();
```
when `WaitingTaskHolder` constructor increases the reference count of the task?

Comment (Dr15Jones) [reply FWCore/Framework/src/Worker.cc:290]: That was a left over from the conversion and isn't needed anymore. On the other hand, it doesn't really hurt either and will be completely replaced when we do the conversion to `tbb::task_group`.

**Comment by makortel** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Comment by silviodonato** [comment]
see https://github.com/cms-sw/cmssw/issues/32563

File: FWCore/Framework/src/Worker.cc [hunk 4]
```diff
@@ -267,63 +268,65 @@ namespace edm {
...
-          make_waiting_task(tbb::task::allocate_root(),
-                            [holder = WaitingTaskWithArenaHolder{iTask}](std::exception_ptr const* iExcept) mutable {
-                              if (iExcept) {
-                                holder.doneWaiting(*iExcept);
-                              } else {
-                                holder.doneWaiting(std::exception_ptr{});
-                              }
-                            });
+      auto task = make_waiting_task(tbb::task::allocate_root(),
+                                    [holder = std::move(iTask)](std::exception_ptr const* iExcept) mutable {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Worker.cc:307]: This is a bug, holder needs to be of type `WaitingTaskForArenaHolder`

**Pull request merged**